internal/acs/runner.go

```diff
@@ -0,0 +1,250 @@
+}
+
+type ACSCredentials struct {
+	Endpoint string
+	Token    string
+}
+type cveSummaryData struct {
+	clusterName     string
+	namespaceName   string
+	instances       map[string]int
+	imageID         string
+	imageRegistry   string
+	imageRemote     string
+	imageTag        string
+	operatingSystem string
+	cveData         map[string]cveData
+}
+type cveData struct {
+	severity  string
+	cvss      float64
+	isFixable bool
+}
+
+func CVECollectorLoop(collectorParams CollectorLoopParams) {
+	acsClient, err := acsclient.NewACSClient(collectorParams.Credentials.Endpoint, collectorParams.Credentials.Token, collectorParams.Timeout, collectorParams.Retry)
+	if err != nil {
+		level.Error(collectorParams.Logger).Log("msg", "error creating acs client", "error", err.Error())
+		collectorParams.ExitSignal <- struct{}{}
+		return
+	}
+	for {
+		cveData, err := getCVEDataFromACS(acsClient, collectorParams.WorkerPoolSize)
+		if err != nil {
+			level.Error(collectorParams.Logger).Log("msg", "Error processing metrics::", "error", err.Error())
+			collectorParams.ExitSignal <- struct{}{}
+			return
+		}
+
+		collectorParams.Collector.MetricsData = getMetricDataFromCVE(cveData, collectorParams.Logger, collectorParams.Collector)
+		level.Debug(collectorParams.Logger).Log("Collection completed.")
+	}
+}
+
+func getCVEDataFromACS(client acsclient.Client, workerPoolSize int) ([]interface{}, error) {
+	clusters, err := client.GetClusters()
+	if err != nil {
+		return nil, fmt.Errorf("error getting clusters:%v", err)
+	}
+	workerPool := worker.NewWorkerPool(workerPoolSize, true)
+	for _, cluster := range clusters {
+		// to avoid goroutine in loop error
+		clusterData := cluster
+		workerPool.Run(func() (interface{}, error) {
+			return processClusterCVEs(clusterData, client)
+		})
+	}
+	return workerPool.WaitAndReturnData()
+}
+
+func getMetricDataFromCVE(records []interface{}, logger log.Logger, collector *collector.CVECollector) map[string]prometheus.Metric {
+	metricMap := make(map[string]prometheus.Metric)
+	for _, record := range records {
+		if cveMap, ok := record.(map[string]cveSummaryData); ok {
+			for _, cveSummaryData := range cveMap {
+				for key, val := range cveSummaryData.cveData {
+					labels := prometheus.Labels{
+						"cluster":         cveSummaryData.clusterName,
+						"cve":             key,
+						"severity":        val.severity,
+						"isFixable":       strconv.FormatBool(val.isFixable),
+						"deploymentCount": strconv.FormatInt(int64(len(cveSummaryData.instances)), 10),
+						"namespace":       cveSummaryData.namespaceName,
+						"imageRemote":     cveSummaryData.imageRemote,
+						"imageRegistry":   cveSummaryData.imageRegistry,
+						"imageTag":        cveSummaryData.imageTag,
+						"imageId":         cveSummaryData.imageID,
+						"operatingSystem": cveSummaryData.operatingSystem,
+					}
+
+					index := ConvertLabelsToString(labels)
+					level.Info(logger).Log("msg", "Ensuring metric", "index", index, "cvss", val.cvss)
```

Comment:
is `index` only used for logging? do we need to use a map for `metricMap`? can it just be a list? any duplicates?

Reply:
index is used as a key for cvss value... map was chosen for a condtion if cvss associated with image changes then updation can be easy...(have not noticed it yet just as assumption)..If a list is used..it will create duplicate records in such case

Reply:
can you give a concrete example of duplicate records? with current logic, last one wins, which type of data can cause duplicate?

Reply:
I do not have any example its just a scenario if we have duplicate..in that case we will have same labels but different cvss value...this will cause prometheus to fail with duplicate labels error. having a map instead of slice prevents this error